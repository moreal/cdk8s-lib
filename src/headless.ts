// import { Deployment, Protocol } from "cdk8s-plus-22";
import { Construct } from "constructs";
import { KubeService, IntOrString, Quantity } from "cdk8s-plus-22/lib/imports/k8s";
import { KubeStatefulSet } from "../imports/k8s";

interface HostAndPort { host: string, port: number };

type LogLevel = "error" | "warning" | "information" | "debug" | "verbose"

export interface HeadlessOptions {
    image: string,
    logLevel: LogLevel,
    workers: number | undefined,
    appProtocolVersion: string,
    libplanet: HostAndPort,
    graphql: HostAndPort | undefined,
    iceServers: string[] | undefined,
    peers: string[] | undefined,
    privateKey: string,
    store: {
        type: string,
        path: string,
    }
    chainTipStaleBehaviour: string,
    minimumBroadcastTarget: number,
};

export class Explorer extends Construct {
    constructor(scope: Construct, id: string, options: HeadlessOptions) {
        super(scope, id);

        const VOLUME_NAME = `${id}-data`;

        const iceServers = options.iceServers || [];
        const iceServersArgs = iceServers.flatMap(iceServer => ["--ice-server", iceServer]);

        const peers = options.peers || [];
        const peersArgs = peers.map(peer => `--peer=${peer}`);

        const graphqlArgs = options.graphql === undefined
            ? []
            : [
                `--graphql-server`,
                `--graphql-host=${options.graphql.host}`,
                `--graphql-port=${options.graphql.port}`
            ];

        new KubeStatefulSet(this, `${id}-statefulset`, {
            metadata: {
                labels: {
                    app: id
                },
                name: id,
            },
            spec: {
                replicas: 1,
                selector: {
                    matchLabels: {
                        app: id,
                    }
                },
                serviceName: id,
                updateStrategy: {
                    type: "OnDelete",
                },
                template: {
                    spec: {
                        containers: [
                            {
                                name: id,
                                image: options.image,
                                command: ["dotnet"],
                                args: [
                                    "NineChronicles.Headless.Executable.dll",
                                    "run",
                                    `--log-level=${options.logLevel}`,
                                    `--app-protocol-version=${options.appProtocolVersion}`,
                                    `--host=${options.libplanet.host}`,
                                    `--port=${options.libplanet.port}`,
                                    ...(options.workers === undefined ? [] : [`--workers=${options.workers}`]),
                                    `--private-key=${options.privateKey}`,
                                    ...graphqlArgs,
                                    ...iceServersArgs,
                                    ...peersArgs,
                                ],
                                ports: [
                                    {
                                        containerPort: options.libplanet.port,
                                        name: "node",
                                        protocol: "TCP",
                                    },
                                    ...(options.graphql !== undefined ?
                                        [{
                                            containerPort: options.graphql.port,
                                            name: "graphql",
                                            protocol: "TCP",
                                        }] : [])
                                ],
                                livenessProbe: {
                                    failureThreshold: 3,
                                    initialDelaySeconds: 120,
                                    periodSeconds: 5,
                                    successThreshold: 1,
                                    tcpSocket: {
                                        port: IntOrString.fromNumber(options.libplanet.port),
                                    },
                                    timeoutSeconds: 1,
                                },
                                resources: {
                                    requests: {
                                        cpu: Quantity.fromString("1300m"),
                                        memory: Quantity.fromString("5Gi"),
                                    }
                                },
                                volumeMounts: [
                                    {
                                        mountPath: "/data",
                                        name: VOLUME_NAME
                                    }
                                ]
                            },
                        ],
                        restartPolicy: "Always",
                        dnsPolicy: "ClusterFirst",
                    }
                },
                volumeClaimTemplates: [
                    {
                        metadata: {
                            name: VOLUME_NAME,
                        },
                        spec: {
                            accessModes: ["ReadWriteOnce"],
                            resources: {
                                requests: {
                                    storage: Quantity.fromString("100Gi"),
                                }
                            },
                            storageClassName: "gp2-extensible",
                            volumeMode: "FileSystem"
                        }
                    }
                ]
            }
        });

        new KubeService(this, `${id}-service`, {
            metadata: {
                name: id,
            },
            spec: {
                selector: {
                    app: id,
                },
                type: "LoadBalancer",
                ports: [
                    {
                        name: "node",
                        port: options.libplanet.port,
                        targetPort: IntOrString.fromNumber(options.libplanet.port),
                    },
                    ...(options.graphql === undefined
                        ? []
                        : [{
                            name: "graphql",
                            port: options.graphql.port,
                            targetPort: IntOrString.fromNumber(options.graphql.port),
                        }]),
                ]
            }
        });
    }
}
