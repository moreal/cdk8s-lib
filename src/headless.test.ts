import { Testing } from 'cdk8s';
import { Headless, PrivateKeyStrategy } from './headless';

describe(Headless.name, () => {
    test('snapshot', () => {
        const chart = Testing.chart();
        new Headless(chart, "main-miner-1", {
            appProtocolVersion: "100080/6ec8E598962F1f475504F82fD5bF3410eAE58B9B/MEUCIQCrSQ9v7NZBo5yCyvlHgBXW6lD4p2lFlKqK+aMeoKA.IAIgVkcK6hMCw1E0mf+m7rCf+9XnSL8Apg.joe6I1a8Nw70=/ZHUxNjpXaW5kb3dzQmluYXJ5VXJsdTU2Omh0dHBzOi8vZG93bmxvYWQubmluZS1jaHJvbmljbGVzLmNvbS92MTAwMDgwL1dpbmRvd3MuemlwdTk6dGltZXN0YW1wdTEwOjIwMjEtMDktMzBl",
            trustedAppProtocolVersionSigner: "03eeedcd574708681afb3f02fb2aef7c643583089267d17af35e978ecaf2a1184e",
            genesisBlockPath: "https://9c-test.s3.ap-northeast-2.amazonaws.com/genesis-block-9c-main",
            libplanet: {
              host: "9c-main-miner-1.planetarium.dev",
              port: 31234,
              open: true,
            },
            store: {
              path: "/data/miner",
              type: "rocksdb",
            },
            swarmPrivateKey: {
              strategy: PrivateKeyStrategy.FromSecret,
              secret: {
                key: "miner-keys",
                name: "miner1",
              },
            },
            minerPrivateKey: {
              strategy: PrivateKeyStrategy.FromSecret,
              secret: {
                key: "miner-keys",
                name: "miner1",
              },
            },
            graphql: {
              host: "0.0.0.0",
              port: 80,
              open: true,
            },
            minimumBroadcastTarget: 50,
            txQuotaPerSigner: 500,
            image: "planetariumhq/ninechronicles-headless:v100080-mt-2021093001",
            replicas: 1,
            peers: [
              "027bd36895d68681290e570692ad3736750ceaab37be402442ffb203967f98f7b6,9c-main-seed-1.planetarium.dev,31234",
              "02f164e3139e53eef2c17e52d99d343b8cbdb09eeed88af46c352b1c8be6329d71,9c-main-seed-2.planetarium.dev,31234",
              "0247e289aa332260b99dfd50e578f779df9e6702d67e50848bb68f3e0737d9b9a5,9c-main-seed-3.planetarium.dev,31234",
            ],
          })
        const results = Testing.synth(chart)
        expect(results).toMatchSnapshot();
    });
});
