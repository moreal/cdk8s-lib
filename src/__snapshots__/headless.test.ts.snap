// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headless full-state snapshot 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": Object {
      "labels": Object {
        "app": "main-full-state",
      },
      "name": "main-full-state",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "main-full-state",
        },
      },
      "serviceName": "main-full-state",
      "template": Object {
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "NineChronicles.Headless.Executable.dll",
                "run",
                "--app-protocol-version=100080/6ec8E598962F1f475504F82fD5bF3410eAE58B9B/MEUCIQCrSQ9v7NZBo5yCyvlHgBXW6lD4p2lFlKqK+aMeoKA.IAIgVkcK6hMCw1E0mf+m7rCf+9XnSL8Apg.joe6I1a8Nw70=/ZHUxNjpXaW5kb3dzQmluYXJ5VXJsdTU2Omh0dHBzOi8vZG93bmxvYWQubmluZS1jaHJvbmljbGVzLmNvbS92MTAwMDgwL1dpbmRvd3MuemlwdTk6dGltZXN0YW1wdTEwOjIwMjEtMDktMzBl",
                "--trusted-app-protocol-version-signer=03eeedcd574708681afb3f02fb2aef7c643583089267d17af35e978ecaf2a1184e",
                "--genesis-block-path=https://9c-test.s3.ap-northeast-2.amazonaws.com/genesis-block-9c-main",
                "--minimum-broadcast-target=50",
                "--tx-quota-per-signer=500",
                "--port=31234",
                "--graphql-server",
                "--graphql-host=0.0.0.0",
                "--graphql-port=80",
                "--peer=027bd36895d68681290e570692ad3736750ceaab37be402442ffb203967f98f7b6,9c-main-seed-1.planetarium.dev,31234",
                "--peer=02f164e3139e53eef2c17e52d99d343b8cbdb09eeed88af46c352b1c8be6329d71,9c-main-seed-2.planetarium.dev,31234",
                "--peer=0247e289aa332260b99dfd50e578f779df9e6702d67e50848bb68f3e0737d9b9a5,9c-main-seed-3.planetarium.dev,31234",
                "--swarm-private-key=$(SWARM_PRIVATE_KEY)",
              ],
              "command": Array [
                "dotnet",
              ],
              "env": Array [
                Object {
                  "name": "SWARM_PRIVATE_KEY",
                  "valueFrom": Object {
                    "secretKeyRef": Object {
                      "key": "miner-keys",
                      "name": "miner1",
                    },
                  },
                },
              ],
              "image": "planetariumhq/ninechronicles-headless:v100080-mt-2021093001",
              "livenessProbe": Object {
                "failureThreshold": 3,
                "initialDelaySeconds": 120,
                "periodSeconds": 5,
                "successThreshold": 1,
                "tcpSocket": Object {
                  "port": 31234,
                },
                "timeoutSeconds": 1,
              },
              "name": "main-full-state",
              "ports": Array [
                Object {
                  "containerPort": 80,
                  "name": "graphql",
                  "protocol": "TCP",
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "1300m",
                  "memory": "5Gi",
                },
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/data",
                  "name": "main-full-state-data",
                },
              ],
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
        },
      },
      "updateStrategy": Object {
        "type": "OnDelete",
      },
      "volumeClaimTemplates": Array [
        Object {
          "metadata": Object {
            "name": "main-full-state-data",
          },
          "spec": Object {
            "accessModes": Array [
              "ReadWriteOnce",
            ],
            "resources": Object {
              "requests": Object {
                "storage": "100Gi",
              },
            },
            "storageClassName": "gp2-extensible",
            "volumeMode": "Filesystem",
          },
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "main-full-state",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "graphql",
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "app": "main-full-state",
      },
      "type": "LoadBalancer",
    },
  },
]
`;

exports[`Headless snapshot 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": Object {
      "labels": Object {
        "app": "main-miner-1",
      },
      "name": "main-miner-1",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "main-miner-1",
        },
      },
      "serviceName": "main-miner-1",
      "template": Object {
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "NineChronicles.Headless.Executable.dll",
                "run",
                "--app-protocol-version=100080/6ec8E598962F1f475504F82fD5bF3410eAE58B9B/MEUCIQCrSQ9v7NZBo5yCyvlHgBXW6lD4p2lFlKqK+aMeoKA.IAIgVkcK6hMCw1E0mf+m7rCf+9XnSL8Apg.joe6I1a8Nw70=/ZHUxNjpXaW5kb3dzQmluYXJ5VXJsdTU2Omh0dHBzOi8vZG93bmxvYWQubmluZS1jaHJvbmljbGVzLmNvbS92MTAwMDgwL1dpbmRvd3MuemlwdTk6dGltZXN0YW1wdTEwOjIwMjEtMDktMzBl",
                "--trusted-app-protocol-version-signer=03eeedcd574708681afb3f02fb2aef7c643583089267d17af35e978ecaf2a1184e",
                "--genesis-block-path=https://9c-test.s3.ap-northeast-2.amazonaws.com/genesis-block-9c-main",
                "--minimum-broadcast-target=50",
                "--tx-quota-per-signer=500",
                "--host=9c-main-miner-1.planetarium.dev",
                "--port=31234",
                "--graphql-server",
                "--graphql-host=0.0.0.0",
                "--graphql-port=80",
                "--peer=027bd36895d68681290e570692ad3736750ceaab37be402442ffb203967f98f7b6,9c-main-seed-1.planetarium.dev,31234",
                "--peer=02f164e3139e53eef2c17e52d99d343b8cbdb09eeed88af46c352b1c8be6329d71,9c-main-seed-2.planetarium.dev,31234",
                "--peer=0247e289aa332260b99dfd50e578f779df9e6702d67e50848bb68f3e0737d9b9a5,9c-main-seed-3.planetarium.dev,31234",
                "--swarm-private-key=$(SWARM_PRIVATE_KEY)",
                "--miner-private-key=$(MINER_PRIVATE_KEY)",
              ],
              "command": Array [
                "dotnet",
              ],
              "env": Array [
                Object {
                  "name": "SWARM_PRIVATE_KEY",
                  "valueFrom": Object {
                    "secretKeyRef": Object {
                      "key": "miner-keys",
                      "name": "miner1",
                    },
                  },
                },
                Object {
                  "name": "MINER_PRIVATE_KEY",
                  "valueFrom": Object {
                    "secretKeyRef": Object {
                      "key": "miner-keys",
                      "name": "miner1",
                    },
                  },
                },
              ],
              "image": "planetariumhq/ninechronicles-headless:v100080-mt-2021093001",
              "livenessProbe": Object {
                "failureThreshold": 3,
                "initialDelaySeconds": 120,
                "periodSeconds": 5,
                "successThreshold": 1,
                "tcpSocket": Object {
                  "port": 31234,
                },
                "timeoutSeconds": 1,
              },
              "name": "main-miner-1",
              "ports": Array [
                Object {
                  "containerPort": 31234,
                  "name": "node",
                  "protocol": "TCP",
                },
                Object {
                  "containerPort": 80,
                  "name": "graphql",
                  "protocol": "TCP",
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "1300m",
                  "memory": "5Gi",
                },
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/data",
                  "name": "main-miner-1-data",
                },
              ],
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
        },
      },
      "updateStrategy": Object {
        "type": "OnDelete",
      },
      "volumeClaimTemplates": Array [
        Object {
          "metadata": Object {
            "name": "main-miner-1-data",
          },
          "spec": Object {
            "accessModes": Array [
              "ReadWriteOnce",
            ],
            "resources": Object {
              "requests": Object {
                "storage": "100Gi",
              },
            },
            "storageClassName": "gp2-extensible",
            "volumeMode": "Filesystem",
          },
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "main-miner-1",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "node",
          "port": 31234,
          "targetPort": 31234,
        },
        Object {
          "name": "graphql",
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "app": "main-miner-1",
      },
      "type": "LoadBalancer",
    },
  },
]
`;
